src/transaction/log_meta.cpp

```diff
@@ -178,6 +178,7 @@ namespace cublog
   void
   meta::add_checkpoint_info (const log_lsa &chkpt_lsa, const checkpoint_info &chkpt_info)
   {
+    assert (m_checkpoints.find (chkpt_lsa) == m_checkpoints.cend ());
```

Comment:
If no log records are added during two checkpoints, it seems that `chkpt_lsa` could be in `m_checkpoints`. This can't happen?

Reply:
`log_Checkpoint_trantable_daemon` is set to execute every 60 seconds (`log_checkpoint_trantable_daemon_init`).
I know that there are status/dummy log records which are generated more often (eg: `LOG_DUMMY_HA_SERVER_STATE`) - which replication also uses to piggyback replication delay statistics calculation on page server and, in the future, passive transaction server.

Reply:
> log_Checkpoint_trantable_daemon is set to execute every 60 seconds (log_checkpoint_trantable_daemon_init).

In ATS, it doesn't seem to be a problem because it generates the tran-table snapshot log record together with adding a checkping info in the meta log, which makes LSA advanced. Where I think it could matter is in PS. 

> know that there are status/dummy log records which are generated more often (eg: LOG_DUMMY_HA_SERVER_STATE) - which replication also uses to piggyback replication delay statistics calculation on page server and, in the future, passive transaction server.
Later edit: Might it be safer to just, if a same LSA checkpoint is found already in the map, replace with the new one?

I suppose it is better to decouple what generates logs and the checkpoint/log mechanism. Your suggestion later edited looks better to me if the situation I mentioned can occur :).

Reply:
Or can we just skip adding checkpoint? No record after the previous checkpoint means no need to do a checkpoint?

Reply:
PS's do not generate log and, I think, do not perform checkpoints.
On ATS, indeed, a log record is added prior to saving the checkpoint in meta log.
To be congruent with this addition of checkpoint as a log record, I replaced the checkpoint in metalog if a same-LSA checkpoint is already found there.

Reply:
I'd like to correct If I have misunderstood. Could you let me know the followings are correct?
1. On PS, no log records are generated, but checkpoint is performed in logpb_checkpoint(), in which checkpoint info is added in the meta log.
2. On ATS, there is no chance to get the same LSA as the argument in add_checkpoint_info() because a log record is added prior to saving the checkpoint in meta log as you said, and it will advance the snapshot lsa of the checkpoint info in the next time.

Reply:
1. You are correct. I was mistaken. Checkpoint daemon does execute on PS.
2. correct again

Reply:
Thank you for the answer.